File: customerapp/src/main/java/com/tokopedia/tkpd/home/wishlist/adapter/factory/WishlistAdapterFactory.java
package com.tokopedia.tkpd.home.wishlist.adapter.factory;

import android.view.View;

import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory;
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder;
import com.tokopedia.tkpd.home.adapter.OnWishlistActionButtonClicked;
import com.tokopedia.tkpd.home.presenter.WishListView;
import com.tokopedia.tkpd.home.wishlist.adapter.viewholder.*;
import com.tokopedia.tkpd.home.wishlist.adapter.viewmodel.*;
import com.tokopedia.tkpd.home.wishlist.analytics.WishlistAnalytics;

import org.jetbrains.annotations.NotNull;

/**
 * Author errysuprayogi on 25,July,2019
 */
public class WishlistAdapterFactory extends BaseAdapterTypeFactory implements WishlistTypeFactory {

    private OnWishlistActionButtonClicked actionButtonClicked;
    private WishListView wishListView;
    private WishlistAnalytics wishlistAnalytics;

    public WishlistAdapterFactory(OnWishlistActionButtonClicked actionButtonClicked,
                                  WishListView wishListView,
                                  WishlistAnalytics wishlistAnalytics) {
        this.actionButtonClicked = actionButtonClicked;
        this.wishListView = wishListView;
        this.wishlistAnalytics = wishlistAnalytics;
    }

    @Override
    public int type(@NotNull WishlistProductViewModel viewModel) {
        return WishlistProductListViewHolder.LAYOUT;
    }

    @Override
    public int type(@NotNull WishlistEmptyViewModel viewModel) {
        return WishlistEmptyViewHolder.LAYOUT;
    }

    @Override
    public int type(@NotNull WishlistEmptySearchViewModel viewModel) {
        return WishlistEmptySearchViewHolder.LAYOUT;
    }

    @Override
    public int type(@NotNull WishlistTopAdsViewModel viewModel) {
        return WishlistTopAdsListViewHolder.LAYOUT;
    }

    @Override
    public int type(@NotNull WishlistRecomendationViewModel viewModel) {
        return WishlistRecomendationViewHolder.LAYOUT;
    }

    @Override
    public int type(@NotNull WishlistRecomTitleViewModel viewModel) {
        return WishlistRecomTitleViewHolder.LAYOUT;
    }

    @Override
    public int type(WishlistRecommendationCarouselViewModel viewModel) {
        return WishlistRecommendationCarouselViewHolder.LAYOUT;
    }

    @Override
    public AbstractViewHolder createViewHolder(View parent, int type) {
        if(type == WishlistEmptySearchViewHolder.LAYOUT) {
            return new WishlistEmptySearchViewHolder(parent, actionButtonClicked);
        } else if (type == WishlistProductListViewHolder.LAYOUT){
            return new WishlistProductListViewHolder(parent, wishlistAnalytics, wishListView);
        } else if (type == WishlistEmptyViewHolder.LAYOUT){
            return new WishlistEmptyViewHolder(parent, actionButtonClicked);
        } else if (type == WishlistTopAdsListViewHolder.LAYOUT){
            return new WishlistTopAdsListViewHolder(parent);
        } else if (type == WishlistRecomendationViewHolder.LAYOUT){
            return new WishlistRecomendationViewHolder(parent, wishlistAnalytics);
        } else if (type == WishlistRecomTitleViewHolder.LAYOUT){
            return new WishlistRecomTitleViewHolder(parent);
        } else if(type == WishlistRecommendationCarouselViewHolder.LAYOUT){
            return new WishlistRecommendationCarouselViewHolder(parent, wishlistAnalytics);
        }
        return super.createViewHolder(parent, type);
    }
}

File: customerapp/src/main/java/com/tokopedia/tkpd/home/wishlist/adapter/factory/WishlistAdapterFactory.kt
package com.tokopedia.tkpd.home.wishlist.adapter.factory

import android.view.View
import com.tokopedia.abstraction.base.view.adapter.Visitable
import com.tokopedia.abstraction.base.view.adapter.factory.BaseAdapterTypeFactory
import com.tokopedia.abstraction.base.view.adapter.viewholders.AbstractViewHolder
import com.tokopedia.tkpd.home.adapter.OnWishlistActionButtonClicked
import com.tokopedia.tkpd.home.wishlist.adapter.viewholder.*
import com.tokopedia.tkpd.home.wishlist.adapter.viewmodel.*
import com.tokopedia.tkpd.home.presenter.WishListView
import com.tokopedia.tkpd.home.wishlist.analytics.WishlistAnalytics

/**
 * Author errysuprayogi on 03,July,2019
 */
class WishlistAdapterFactory(val buttonActionClick: OnWishlistActionButtonClicked, val wishListView: WishListView, val wishlistAnalytics: WishlistAnalytics) :
        BaseAdapterTypeFactory(), WishlistTypeFactory {

    override fun type(viewModel: WishlistProductViewModel): Int {
        return WishlistProductListViewHolder.LAYOUT
    }

    override fun type(viewModel: WishlistEmptyViewModel): Int {
        return WishlistEmptyViewHolder.LAYOUT
    }

    override fun type(viewModel: WishlistEmptySearchViewModel): Int {
        return WishlistEmptySearchViewHolder.LAYOUT
    }

    override fun type(viewModel: WishlistTopAdsViewModel): Int {
        return WishlistTopAdsListViewHolder.LAYOUT
    }

    override fun type(viewModel: WishlistRecomendationViewModel): Int {
        return WishlistRecomendationViewHolder.LAYOUT
    }

    override fun type(viewModel: WishlistRecomTitleViewModel): Int {
        return WishlistRecomTitleViewHolder.LAYOUT
    }

    override fun createViewHolder(view: View, viewType: Int): AbstractViewHolder<out Visitable<*>> {
        return when (viewType) {
            WishlistProductListViewHolder.LAYOUT -> WishlistProductListViewHolder(view, wishlistAnalytics, wishListView)
            WishlistEmptyViewHolder.LAYOUT -> WishlistEmptyViewHolder(view, buttonActionClick)
            WishlistEmptySearchViewHolder.LAYOUT -> WishlistEmptySearchViewHolder(view, buttonActionClick)
            WishlistTopAdsListViewHolder.LAYOUT -> WishlistTopAdsListViewHolder(view)
            WishlistRecomendationViewHolder.LAYOUT -> WishlistRecomendationViewHolder(view)
            WishlistRecomTitleViewHolder.LAYOUT -> WishlistRecomTitleViewHolder(view)
            else -> super.createViewHolder(view, viewType)
        }
    }
}
